Add pre-commit config & associated instructions

yongrenjie opened the pull request:
~~**Currently, more checks are required to ensure that `pre-commit` sets everything up as required; as per @llewelld's comments below this doesn't seem to actually be the case.** In particular it seems to prefer using the macOS built-in `ruby` rather than installing its own copy.~~

## Summary

Added the following:

 - A note in the README linking to the contributing page in the handbook itself. (I think people often look there first.)
 - `.pre-commit-config.yaml`, which (almost) follows the same steps as in `ci.yaml`, except for the html-proofer.
 - Instructions on how to perform one-off Markdown linting, either using `pre-commit` or manually.
 - Instructions on setting up pre-commit hooks.


## Related issues

- Closes #69.
- Partially fulfills #28, except for the html-proofer. Unfortunately, [that repo](https://github.com/gjtorikian/html-proofer) doesn't expose a `.pre-commit-hooks.yaml` so the only way to add a pre-commit hook for that would be to:
  - submit a PR to that repo adding one
  - add a local hook to run the command, but that would require a local installation of html-proofer.

llewelld commented on content/docs/contributing/contributing_changes.md:
```diff
@@ -29,133 +27,148 @@ In short
-review. Otherwise, you might end up having to make many changes during review,
-have difficultly keeping your branch up to date or upset the reviewers.  {{%
-/hint %}}
+If you want to manually perform these checks at any point in time, use:
```
I think it might be worth mentioning that -- if I understand correctly -- this will run the checks over files that have changes committed or staged (rather than running it over everything).

llewelld commented on content/docs/contributing/contributing_changes.md:
````diff
@@ -29,133 +27,148 @@ In short
+If you want to manually perform these checks at any point in time, use:
+
+```bash
+pre-commit run
````
In practice, when I tried following these steps I ended up with errors, presumably because I only had the default Ruby installed:
```
An unexpected error has occurred: CalledProcessError: command: ('/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby', '/usr/bin/gem', 'install', '--no-document', '--no-format-executable', '--no-user-install', 'mdl-0.12.0.gem')
return code: 1
stdout: (none)
stderr:
    ERROR:  Error installing mdl-0.12.0.gem:
        There are no versions of mdl (= 0.12.0) compatible with your Ruby & RubyGems
        mdl requires Ruby version >= 2.7. The current ruby version is 2.6.10.210.
Check the log at /Users/dllewellyn-jones/.cache/pre-commit/pre-commit.log```
```
To get it to work I had to do the following (probably there's a nicer way):
```
brew install ruby
PATH="$(brw --prefix ruby)/bin:$PATH" pre-commit run
```
Is there any way to capture this to avoid tripping people up, without making the instructions overly complex?

yongrenjie replied:
Hmmm, that's not good. I was under the impression that `pre-commit` would reinstall its own copy of ruby and mdl. (The docs certainly seemed to suggest that [it installs its own copy of Node](https://pre-commit.com/#usage), for example, so I assumed the same was true for other languages.)

I'll investigate more (probably after Away Day)...

Thanks for giving it a go! It definitely needed it! I'll mark this as draft in the meantime I think.

yongrenjie replied:
(I think you can do `brew link ruby --force --overwrite` which should create an appropriate symlink, but `brew` itself does suggest that editing `$PATH` is preferable to that, so I'm not sure how much of a good idea that is.)

llewelld commented on content/docs/contributing/style_guide.md:
```diff
@@ -32,23 +32,53 @@ When it is advantageous, data should be [stored in data files and processed usin
+`pre-commit` is intended to be run before all commits (to make sure that you don't unintentionally commit something unwanted).
+To see how to set this up, see the [Contributing Changes]({{< ref "/content/docs/contributing/contributing_changes.md#making-changes" >}}) page.
+
+### Manual Installation
```
I'd personally suggest removing this **Manual Installation** section entirely now. My feeling is that it adds complexity, without adding much benefit given that `pre-commit run markdownlint --all-files` achieves much the same.

yongrenjie replied:
I was thinking about that, too.

yongrenjie replied:
(But let's wait until the `pre-commit` setup is definitely sorted out. 😄)

llewelld commented on content/docs/contributing/contributing_changes.md:
```diff
@@ -29,133 +27,148 @@ In short
-don't want your changes to be merged yet but would like some feedback, you can
-open a draft pull request.
+When you think your changes are ready to be merged, open a pull request.
+If you don't want your changes to be merged yet but would like some feedback, you can open a draft pull request.
```
It's great that you've reformatted the lines at the same time :smiley:

llewelld commented on content/docs/contributing/style_guide.md:
```diff
@@ -32,23 +32,53 @@ When it is advantageous, data should be [stored in data files and processed usin
 
-Markdown styling is enforced by [markdownlint](https://github.com/markdownlint/markdownlint) using the configuration {{% repo_link path=".mdl_style.rb" text=".mdl_style.rb" %}}.
-An explanation of the rules can be found [here](https://github.com/markdownlint/markdownlint/blob/master/docs/RULES.md).
+Markdown styling is enforced by [`markdownlint`](https://github.com/markdownlint/markdownlint) using the configuration {{% repo_link path=".mdl_style.rb" text=".mdl_style.rb" %}}.
```
Nice that you've formatting the `markdownlint` instances too.

llewelld reviewed (commented):
Really nice changes. I've made a few small suggestions, but especially the pre-commit hooks make a really nice addition.

Structurally, the linting info is split across the **Style Guide** and **Contributing Changes** pages (this isn't a new thing, it was like that before) and it feels to me it would be cleaner if it was all on one page. Not something for this PR though.

JimMadge commented on content/docs/contributing/style_guide.md:
```diff
@@ -32,23 +32,53 @@ When it is advantageous, data should be [stored in data files and processed usin
 
-Once installed and assuming you're in the repository root, you can lint a specific file you've made changes to like this:
+First, install Ruby with Homebrew, and then use that to install `mdl`.
+macOS actually provides a system installation of Ruby, but using that isn't advisable as you need to install gems as the root user.
```
I'm not sure if this is true. Using the system `gem` packages I install as a normal user go to `$GEM_HOME`.

yongrenjie replied:
As above I think we might just delete this block, so the rest is a bit of a digression (feel free to ignore). I've almost no experience with ruby so I might be wrong, but this was my experience with system `ruby` and `gem`.

@llewelld Sorry, I can't remember — did you mention you managed to install `mdl` with system `gem`?

```bash
$ which -a gem
/opt/homebrew/opt/ruby/bin/gem
/usr/bin/gem

$ /usr/bin/gem install mdl
Fetching chef-utils-18.1.29.gem
ERROR:  While executing gem ... (Gem::FilePermissionError)
    You don't have write permissions for the /Library/Ruby/Gems/2.6.0 directory.

$ echo $GEM_HOME


$ sudo /usr/bin/gem install mdl
Password:
Fetching concurrent-ruby-1.2.2.gem
Fetching mixlib-cli-2.1.8.gem
Fetching tomlrb-2.0.3.gem
Fetching mixlib-config-3.0.27.gem
Fetching chef-utils-18.1.29.gem
Fetching kramdown-2.4.0.gem
Fetching mixlib-shellout-3.2.7.gem
Fetching kramdown-parser-gfm-1.1.0.gem
Fetching mdl-0.12.0.gem
Successfully installed kramdown-2.4.0
Successfully installed kramdown-parser-gfm-1.1.0
Successfully installed mixlib-cli-2.1.8
Successfully installed tomlrb-2.0.3
Successfully installed mixlib-config-3.0.27
Successfully installed concurrent-ruby-1.2.2
Successfully installed chef-utils-18.1.29
Successfully installed mixlib-shellout-3.2.7
ERROR:  Error installing mdl:
	The last version of mdl (>= 0) to support your Ruby & RubyGems was 0.11.0. Try installing it with `gem install mdl -v 0.11.0`
	mdl requires Ruby version >= 2.7. The current ruby version is 2.6.10.210.
```

yongrenjie commented on .mdlrc:
```diff
@@ -1,2 +0,0 @@
-style '.mdl_style.rb'
-ignore_front_matter true
```
> This could break things for people not using pre-commit, or calling `mdl` from their editor.

That's true. I removed this because (as I understand it) these two invocations of mdl are the same:

 - (option 1) `mdl -c .mdlrc`
 - (option 2) `mdl -i -s .mdl_style.rb`

Because the pre-commit configuration, as well as CI, use option 2, neither of them require `.mdlrc`. So, there's some information duplication going on here — suppose we were to ever modify `.mdlrc`, it'd be easier to forget to change the command-line arguments (option 2) wherever they appear.

Actually, maybe a better way is to stop explicitly specifying the command-line arguments (option 2) and just stick to the config file (option 1) everywhere?

JimMadge replied:
> Actually, maybe a better way is to stop explicitly specifying the command-line arguments (option 2) and just stick to the config file (option 1) everywhere?

That sounds like a good idea, less typing and duplication.

I _think_ `mdl` looks for `.mdlrc` in the current directory also.

yongrenjie replied:
> I think `mdl` looks for `.mdlrc` in the current directory also.

Yes, it does; so we don't actually need any argument at all.

(I found it can search upwards in the directory structure too, so it can find `.mdlrc` when called from a subdirectory, but sadly it doesn't then find `.mdl_style.rb`.)

yongrenjie replied:
Okay, this should be resolved now. I restored `.mdlrc` to its former glory (:)) and `mdl` should pick it up every time it's run (either manually like in CI, or via pre-commit).

yongrenjie commented on content/docs/contributing/style_guide.md:
````diff
@@ -32,23 +32,53 @@ When it is advantageous, data should be [stored in data files and processed usin
+The easiest way to do this is to use the [pre-commit](https://pre-commit.com/) tool, which has already been set up in the handbook repository.
+
+```bash
+brew install pre-commit   # if not already installed
````
Points taken! Yes, we could remove it to be more agnostic.

I think I was going off this, to which I'm not sure if the same points apply:

https://github.com/alan-turing-institute/REG-handbook/blob/7a22cd13917aeb50c853d08caa46c75e1bd7a1a6/content/docs/onboarding/new_joiners/systems_set_up.md?plain=1#L74-L80

(I suppose the Slack part is targeted at a reader, whereas the contributing section is targeted at a contributor. In principle, at least, they are different.)

JimMadge replied:
I think that difference probably wasn't intentional, although it does make sense! The new joiners guide will talk about Mac stuff as, at the moment, everyone gets a MacBook.

In the contribution guide I tried to rely on external documentation and not make too many assuptions about what tools people will use.

yongrenjie commented on content/docs/contributing/style_guide.md:
````diff
@@ -32,23 +32,53 @@ When it is advantageous, data should be [stored in data files and processed usin
 {{% /hint %}}
 
-Once installed and assuming you're in the repository root, you can lint a specific file you've made changes to like this:
+First, install Ruby with Homebrew, and then use that to install `mdl`.
+macOS actually provides a system installation of Ruby, but using that isn't advisable as you need to install gems as the root user.
+
+```bash
+brew install ruby
+$(brew --prefix ruby)/bin/gem install mdl
+```
+
+Now, run
+
+```bash
+$(brew --prefix ruby)/bin/gem env | grep 'EXECUTABLE DIRECTORY'
+```
````
We might just remove the whole thing.

yongrenjie commented on mlc_config.json:
```diff
@@ -17,6 +17,12 @@
     },
     {
       "pattern": "^https://docs.github.com"
+    },
+    {
+      "pattern": "^https://thealanturininstitute.sharepoint.com"
+    },
+    {
+      "pattern": "^https://wimbledon-planner.azurewebsites.net"
```
On my end they always fail with a 401 status code. I looked into this for a while the other day, but I'm honestly not entirely sure why this check always passes in CI.

JimMadge replied:
They pass in CI? Looking at the addresses, I wouldn't be surprised if they got 401 in CI as they probably authenticate using the Turing AD.

yongrenjie replied:
I'm not sure how or why, but CI somehow manages a 200 for both sites: https://github.com/alan-turing-institute/REG-handbook/actions/runs/4406189381/jobs/7718076564

They always fail locally, though, so I think it's better to exclude them (especially if someone's going to be using pre-commit).

dsj976 commented:
What happened with this pull request finally? Now that I am updating the Handbook I have had similar issues with the markdown lint GitHub Action failing, and I was actually working on fixing the pre-commit configuration before I saw this pull request. I think it would be useful to merge it (at least the updates to `.pre-commit-config.yaml`)

dsj976 commented:
Also, from what I can see here, the `markdownlint` pre-commit hook only flags issues, but does not fix them. I think using [markdownlint-cli](https://github.com/igorshubovych/markdownlint-cli/tree/master) with the `--fix` flag might be of help? No need to add it as a pre-commit hook, can be installed locally and run manually to fix the some of the issues flagged.

yongrenjie commented:
@dsj976 I guess I must've gotten distracted with other things. 😅

I think @llewelld and I concurred that we should just get rid of the 'manual installation' section. I can update that to clean stuff up after Hack Week (or maybe during, if I can get a spare 20 mins somewhere!) and then maybe we can get it merged 🙂

dsj976 commented:
Hi @yongrenjie sounds good! Please let me know when you do

yongrenjie commented:
CI actually failed on Twitter!

```
ERROR: 1 dead links found!
[✖] https://twitter.com/turinghut23 → Status: 0
```

Reproducible via cURL:

```bash
$ curl -v https://twitter.com/turinghut23
[...]
> GET /turinghut23 HTTP/2
> Host: twitter.com
> user-agent: curl/7.88.1
> accept: */*
>
< HTTP/2 302
< date: Mon, 19 Jun 2023 16:04:51 GMT
< perf: 7626143928
< vary: Accept
< server: tsa_f
< location: /turinghut23
< set-cookie: guest_id=v1%3A168719069114643344; Max-Age=34214400; Expires=Fri, 19 Jul 2024 16:04:51 GMT; Path=/; Domain=.twitter.com; Secure; SameSite=None
< content-type: text/plain; charset=utf-8
< x-powered-by: Express
< cache-control: no-cache, no-store, max-age=0
< content-length: 34
< x-transaction-id: 38cae486afb2acc0
< strict-transport-security: max-age=631138519
< x-response-time: 109
< x-connection-hash: c768aed08d595ef0f5b14319f34403f7bbbf6edb401e1cad2be615dc04866d79
<
* Connection #0 to host twitter.com left intact
Found. Redirecting to /turinghut23
```

yongrenjie commented:
Okay, sorry I forgot about this for quite a while. I've made some changes and I think it's almost ready for review. There are some other issues, like Twitter failing on CI (comment above), but that can be part of another PR, this one is big enough as it is.

To be clear, I've removed all duplicated info from the [Style Guide page](https://alan-turing-institute.github.io/REG-handbook/docs/contributing/style_guide/), preferring to instead place them in the relevant sections of the [Contributing Changes page](https://alan-turing-institute.github.io/REG-handbook/docs/contributing/contributing_changes/). I also removed the 'manual installation' instructions, since this doesn't seem to be necessary.

There is one other issue which remains, which is to do with the fact that pre-commit doesn't roll its own version of ruby with it, and system ruby is too old to run the latest version of `mdl`. (To reproduce the output below, I had to uninstall all other versions of ruby, then run `pre-commit clean` to clear the cache.)

```bash
$ which ruby
/usr/bin/ruby
$ ruby --version
ruby 2.6.10p210 (2022-04-12 revision 67958) [universal.arm64e-darwin22]
$ pre-commit run -a
[...]
    ERROR:  Error installing mdl-0.12.0.gem:
    	mdl requires Ruby version >= 2.7. The current ruby version is 2.6.10.210.
```

I see a couple of possible ways around this:

1. **Tell people to install a newer version of ruby and place it in their `$PATH`.** For example, I could run `brew install ruby` and then `pre-commit run -a` ran without problems. (I didn't have to muck around with `gem` or anything, it seems that all that got sorted out just fine.)

2. **Downgrade the version of `mdl`.** If I change `.pre-commit-config.yaml` to point to `v0.11.0` of markdownlint, system ruby works just fine, and we don't need to add any extra instructions.

   This would mean losing out on some features (see [changelog](https://github.com/markdownlint/markdownlint/blob/main/CHANGELOG.md)), but I'm not sure we are using any of these (and if we do need them then we can fix that at a later point).

I'm leaning towards (2) myself, but thoughts are welcome! I think I've addressed all the other comments, so this should hopefully be the last blocker for merging this PR.

JimMadge commented on .pre-commit-config.yaml:
```diff
@@ -1,20 +1,30 @@
+        language: ruby
+        types: ["markdown"]
+
+  - repo: https://github.com/tcort/markdown-link-check
+    rev: v3.10.3
+    hooks:
+      - id: markdown-link-check
+        name: Markdown Link Check
+        description: Check links in newly changed Markdown files
+        entry: markdown-link-check
+        args: ["--config", "mlc_config.json"]
+        language: node
+        types: ["markdown"]
```
We shouldn't let external links block commits (or merges), they will fail sometimes due to websites being down, rate limiting, DOS protection.

If we can change this to _only_ check internal links that would be worthwhile, otherwise it can probably be dropped all together.

(From CI too probably. I've found [Lychee](https://lychee.cli.rs/) recently and much prefer it to `html-proofer` so should make that change at some point)

yongrenjie replied:
I think, actually, it only checks external links. This is because almost all of our internal links are of the form

`{{< relref "/docs/onboarding/new_joiners/systems_set_up.md" >}}`

or

`{{< ref "/content/docs/contributing/contributing_changes.md#continuous-integration" >}}`

and it only looks for links which look like `[text](link)`.

So maybe just drop the whole thing?

JimMadge commented on .github/workflows/ci.yaml:
```diff
@@ -21,7 +21,7 @@ jobs:
 
       - name: Lint markdown
-        run: mdl --ignore-front-matter --style .mdl_style.rb ./
+        run: mdl ./
```
Does this now ignore the style?

In any case, if we are using pre-commit then the CI jobs should be `pre-commit run -a markdown-lint`, `pre-commit run -a hugo` (plus any extra jobs). Otherwise the mismatch between pre-commit and CI will just cause confusion.

yongrenjie replied:
> Does this now ignore the style?

Nope, it still respects it! `mdl` will pick up `.mdlrc` from the repository root, which in turn specifies the command-line options that were removed.

> In any case, if we are using pre-commit then the CI jobs should be [...]

That sounds brilliant, let's do that.

JimMadge commented on content/docs/contributing/contributing_changes.md:
```diff
@@ -29,133 +27,151 @@ In short
+
+- Markdown linting, i.e. style checking
+- a Hugo build, to make sure that the pages can be built
+- a Markdown link check
```
```suggestion
```

JimMadge reviewed (changes requested):
Looking good.

Let's make sure the CI matches pre-commit to avoid confusion.
Dropping link checking for pre-commit should save a lot of headaches.

yongrenjie commented:
Okay! **pre-commit** now only has two stages: markdownlint, plus a hugo build - no more link checker. (markdownlint now uses v0.11.0 for reasons described in my previous comment.)

**CI** now does the aforementioned pre-commit checks, plus html-proofer.

I think this PR should be limited to these changes. There are a couple of things we might want to subsequently fix (which I can create issues for):

1. The new CI actually [fails right now](https://github.com/alan-turing-institute/REG-handbook/actions/runs/5323345550/jobs/9641039164), because the following files don't pass markdownlint:

  - `.github/pull_request_template.md`
  - `.github/ISSUE_TEMPLATE/bug_report.md`
  - `.github/ISSUE_TEMPLATE/chapter_proposal.md`
  - `.github/ISSUE_TEMPLATE/chapter_proposal.md`

2. Furthermore, going forward, we may want to:

   - replace `html-proofer` with `lychee`
   - add that to the `pre-commit` config (and make CI use the pre-commit step), for full consistency.

JimMadge reviewed (approved):
Excluding `.github` from markdown lint checks as it was previously and we shouldn't merge something which will cause other PRs to fail.

yongrenjie commented:
Thanks @JimMadge for putting up with the incessant revisions :smile:
